pkg/apis/pipeline/v1beta1/pipelinerun_conversion.go

```diff
@@ -208,6 +216,159 @@ func (ptrs *PipelineTaskRunSpec) convertFrom(ctx context.Context, source v1.Pipe
 }
 
+// ConvertTo implements apis.Convertible
+func (prs *PipelineRunStatus) ConvertTo(ctx context.Context, sink *v1.PipelineRunStatus) error {
+	// TODO: handle ducktype conversion using knative func once https://github.com/knative/pkg/issues/2655 is fixed
+	sink.Status.ObservedGeneration = prs.Status.ObservedGeneration
+	sink.Status.Conditions = duckv1.Conditions(prs.Status.Conditions)
+	sink.Status.Annotations = prs.Status.Annotations
+
+	sink.StartTime = prs.StartTime
+	sink.CompletionTime = prs.CompletionTime
+	sink.Results = nil
+	for _, pr := range prs.PipelineResults {
+		new := v1.PipelineRunResult{}
+		pr.convertTo(ctx, &new)
+		sink.Results = append(sink.Results, new)
+	}
+	if prs.PipelineSpec != nil {
+		sink.PipelineSpec = &v1.PipelineSpec{}
+		err := prs.PipelineSpec.ConvertTo(ctx, sink.PipelineSpec)
+		if err != nil {
+			return err
+		}
+	}
+	sink.SkippedTasks = nil
+	for _, st := range prs.SkippedTasks {
+		new := v1.SkippedTask{}
+		st.convertTo(ctx, &new)
+		sink.SkippedTasks = append(sink.SkippedTasks, new)
+	}
+	sink.ChildReferences = nil
+	for _, cr := range prs.ChildReferences {
+		new := v1.ChildStatusReference{}
+		cr.convertTo(ctx, &new)
+		sink.ChildReferences = append(sink.ChildReferences, new)
+	}
+	sink.FinallyStartTime = prs.FinallyStartTime
+	if prs.Provenance != nil {
+		new := v1.Provenance{}
+		prs.Provenance.convertTo(ctx, &new)
+		sink.Provenance = &new
+	}
+	if prs.TaskRuns != nil {
+		sink.ChildReferences = append(sink.ChildReferences, convertTaskRunsToChildReference(ctx, prs.TaskRuns)...)
+	}
+	if prs.Runs != nil {
+		sink.ChildReferences = append(sink.ChildReferences, convertRunsToChildReference(ctx, prs.Runs)...)
+	}
+	return nil
+}
+
+// ConvertFrom implements apis.Convertible
+func (prs *PipelineRunStatus) ConvertFrom(ctx context.Context, source *v1.PipelineRunStatus) error {
+	// TODO: handle ducktype conversion using knative func once https://github.com/knative/pkg/issues/2655 is fixed
+	prs.Status.ObservedGeneration = source.Status.ObservedGeneration
+	prs.Status.Conditions = v1beta1.Conditions(source.Conditions)
+	prs.Annotations = source.Annotations
+
+	prs.StartTime = source.StartTime
+	prs.CompletionTime = source.CompletionTime
+	prs.PipelineResults = nil
+	for _, pr := range source.Results {
+		new := PipelineRunResult{}
+		new.convertFrom(ctx, pr)
+		prs.PipelineResults = append(prs.PipelineResults, new)
+	}
+	if source.PipelineSpec != nil {
+		newPipelineSpec := PipelineSpec{}
+		err := newPipelineSpec.ConvertFrom(ctx, source.PipelineSpec)
+		if err != nil {
+			return err
+		}
+		prs.PipelineSpec = &newPipelineSpec
+	}
+	prs.SkippedTasks = nil
+	for _, st := range source.SkippedTasks {
+		new := SkippedTask{}
+		new.convertFrom(ctx, st)
+		prs.SkippedTasks = append(prs.SkippedTasks, new)
+	}
+	prs.ChildReferences = nil
+	for _, cr := range source.ChildReferences {
+		new := ChildStatusReference{}
+		new.convertFrom(ctx, cr)
+		prs.ChildReferences = append(prs.ChildReferences, new)
+	}
```

Comment:
I think we'd want to convert to TaskRuns/Runs if the feature flag specifies this, but unfortunately there's not enough information in ChildReferences to populate those fields without making api calls (which we don't want to do in a webhook). I'm concerned the PipelineRun controller might not function correctly if we don't populate the correct section of PR status. We can remove TaskRuns and Runs as early as January (and I think we should) but not sure what to do until then... @abayer wdyt?

Reply:
Thanks @lbernick .
https://github.com/tektoncd/pipeline/issues/5813 is created to track this down and coming up with a solution with testing and evidence.

Reply:
Thanks for the pointers moving forward, put the TODO at #L292

Reply:
Currently addressed this comment, but would love to have more inputs here @abayer @vdemeester :pray:

Reply:
My preferred approach is before merging this PR, to fail conversion if the embedded status is set to "full" or "both". the TODO should be for supporting conversion if the embedded status is set to either of those values.

My reasoning is that allowing conversion for the wrong value of embedded status might lead to confusing bugs in the PipelineRun controller, and I'm not sure yet whether this is the case. Failing conversion will at least make it clearer to devs what the problem is.

Reply:
Yeah, failing seems reasonable to me. We're removing full/both embedded status in the near future (earliest possible date for switching the default from `full` to `minimal` is Jan 25, 2023, and then we can remove the flag and support for `full` and `both` completely a release later), so this isn't going to be a long-lived problem.

Reply:
Thanks @lbernick and @abayer . Updated.
Sorry for mistaking the failing as a follow-up option in the issue.

Reply:
Just to confirm this error handling would not require a release note.

Reply:
As a side note, having a mix of different embedded status in `PipelineRuns` on the sane cluster is possible today.
A user can use full status, and then switch config to minimal, which will lead to both kind of `PipelineRun` to exist in the cluster.
The controller needs to be able to handle those `PipelineRuns` regardless of the current status of the flag.
That said, I cannot think of anything that the `PipelineRun` controller would have to do with the status of completed `PipelineRuns`, so perhaps the only issue could be with `PipelineRuns` that are running when the configuration change happens. We probably don't test that scenario.

When doing conversions we should consider the current setting of the "embedded-status" flag at all, and provide full convertion back and forth between versions.
For the full status specifically, a way to address that could be to use annotations like we do for `PipelineResources` for instance.
Like `PipelineResources`, full status is something that is going away rather soon, but we are stuck with supporting it for a little while longer.

If we decide instead to drop full-status in the v1beta -> v1 convertion (which would be fine by me) we should document that in the release notes and in the v1beta -> v1 migration docs, so that users may know what to expect.

Reply:
Thanks @afrittoli for the conversation and the clarifications offline as well which are very helpful pinpointing the concerns moving forward. I have noted down the concerns about the knative way and controllers’ behaviours in the issue #5813 and the investigation would be done within the time frame of that , which is when swapping the storage version from v1beta1 to v1.

Reply:
> When doing conversions we should not consider the current setting of the "embedded-status" flag at all, and provide full convertion back and forth between versions.

+1 - Conversion should not cause loss of information.

> If we decide instead to drop full-status in the v1beta -> v1 convertion (which would be fine by me) we should document that in the release notes and in the v1beta -> v1 migration docs, so that users may know what to expect.

This sounds good to me!

Reply:
And meanwhile for this PR, currently while the storage version is still in `v1beta1` and we might want to provide users with a headup when converting down from v1 to v1beta1 with `embedded-status` not set to `minimal`. And I will make sure to have the test cases for changing the `embedded-status` as in a followup PR addressing #5813

Reply:
Updated the release note and the respective migration doc change is initiated at https://github.com/tektoncd/pipeline/pull/5827.